### ISSUE TITLE
add joint degree construction algorithm?

### ISSUE THREAD

**Issue by mgjoka** [networkx/networkx#1996]
Hello,

I am a researcher at the University of California, Irvine and I would like to contribute to the Networkx project by writing a new graph construction algorithm that we have recently published in a peer-reviewed Computer Science conference. More specifically, the graph construction algorithm receives as input a  realizable Joint Degree Matrix (JDM) and **provably** always  produces a simple undirected graph that has exactly the input JDM. Our paper titled "Construction of Simple Graphs with a Target Joint Degree Matrix and Beyond" was published in [IEEE Infocom 2015](http://infocom2015.ieee-infocom.org/) and is [freely available here](http://minasgjoka.com/papers/mgjoka_infocom15.pdf)

Conveniently, Networkx already has a function to calculate the Joint degree Matrix using either of the functions degree_mixing_matrix or degree_mixing_dict 
e.g. 

```
>>> import networkx
>>> g = networkx.Graph([(1,2),(1,3),(2,3),(1,4),(2,5),(2,6),(4,6)])
>>> _joint_degree_dict_ = networkx.degree_mixing_dict(g,normalized=False)
>>> print(_joint_degree_dict_)
{1: {4: 1}, 2: {2: 2, 3: 2, 4: 2}, 3: {2: 2, 4: 1}, 4: {1: 1, 2: 2, 3: 1}}
```

As it can be seen, the Joint Degree describes the number of edges between nodes of degree "k" and degree "l" in the graph. For example in graph "g" above, there is 1 edge between nodes of degree 1 and 4, 2 edges between nodes of degree 2 and 2, 2 edges between nodes of degree 2 and 3, 2 edges between nodes of degree 2 and 4 and so on.

The proposed algorithm, named **joint_degree_simple_graph**, takes as input _joint_degree_dict_, then assesses whether it is realizable, and finally constructs a random instance of a simple undirected graph that has exactly the input joint_degree_dict:

```
>>> g_constructed = joint_degree_simple_graph(_joint_degree_dict_)
```

Would the Networkx community be interested in the addition of such a construction algorithm? 
If yes, please advise where would the most appropriate file to add it. Is it perhaps networkx/networkx/generators/degree_seq.py ? Also, any other advice for a new contributor to the Networkx project would be welcome.

Thank you
Minas Gjoka

**Comment by hagberg** [comment]
Hi - thanks for this suggestion.  Your proposal is good - go ahead and add the algorithm.  You might want to just create a new file since degree_seq.py is getting kind of large and random_graphs.py is even bigger.  Maybe joint_degree_seq.py?

**Comment by mgjoka** [comment]
Sounds good about creating a new file.  

Should I attach the file here when I have it ready?

**Comment by hagberg** [comment]
The best way is to make a pull request to the github repository.  That way we can easily make comments and suggestions and help you get the code integrated smoothly into networkx.    There is some info here on how to get started https://networkx.readthedocs.org/en/stable/developer/index.html
Documentation and testing are an important part of a successful contribution so consider writing that while you develop the code.

**Comment by mgjoka** [comment]
I followed the directions in your link (Git for development) and I committed files that implement the proposed feature. I am waiting for someone to review the code and either give me feedback or approve the commit.

**Comment by jfinkels** [comment]
Fixed in pull request #2011.
